gusto/linear_solvers.py

```diff
@@ -220,6 +220,182 @@ def solve(self):
+        dt = state.timestepping.dt
+        beta = dt*state.timestepping.alpha
+        cp = state.parameters.cp
+        mu = state.mu
+        Vu = state.spaces("HDiv")
+        Vu_broken = FunctionSpace(BrokenElement(Vu.ufl_element()))
+        Vtheta = state.spaces("HDiv_v")
+        Vrho = state.spaces("DG")
+        HOW DO WE GET THE TRACE SPACE IN THE EXTRUDED CASE?
+        Vtrace = FunctionSpace(mesh, "HDiv Trace", tdegree)
+
+        # Split up the rhs vector (symbolically)
+        u_in, rho_in, theta_in = split(state.xrhs)
+
+        # Build the reduced function space for u,rho
+        M = MixedFunctionSpace((Vu_broken, Vrho, Vtrace))
+        w, phi, dl = TestFunctions(M)
+        u, rho, l0 = TrialFunctions(M)
+
+        n = FacetNormal(state.mesh)
+
+        # Get background fields
+        thetabar = state.fields("thetabar")
+        rhobar = state.fields("rhobar")
+        pibar = exner(thetabar, rhobar, state)
+        pibar_rho = exner_rho(thetabar, rhobar, state)
+        pibar_theta = exner_theta(thetabar, rhobar, state)
+
+        # Analytical (approximate) elimination of theta
+        k = state.k             # Upward pointing unit vector
+        theta = -dot(k,u)*dot(k,grad(thetabar))*beta + theta_in
+
+        # Only include theta' (rather than pi') in the vertical
+        # component of the gradient
+
+        # the pi prime term (here, bars are for mean and no bars are
+        # for linear perturbations)
+
+        pi = pibar_theta*theta + pibar_rho*rho
+
+        # vertical projection
+        def V(u):
+            return k*inner(u,k)
+
+        # specify degree for some terms as estimated degree is too large
+        dxp = dx(degree=(self.quadrature_degree))
+        dS_vp = dS_v(degree=(self.quadrature_degree))
+        dS_hp = dS_v(degree=(self.quadrature_degree))
+
+        eqn = (
+            inner(w, (u - u_in))*dx
+            - beta*cp*div(theta*V(w))*pibar*dxp
+            # following does nothing but is preserved in the comments
+            # to remind us why (because V(w) is purely vertical.
+            # + beta*cp*jump(theta*V(w),n)*avg(pibar)*dS_v
+            - beta*cp*div(thetabar*w)*pi*dxp
+            + beta*cp*jump(thetabar*w,n)*l0*(dS_vp + dS_hp)
+            + (phi*(rho - rho_in) - beta*inner(grad(phi), u)*rhobar)*dx
+            + beta*jump(phi*u, n)*avg(rhobar)*(dS_v + dS_h)
+            + dl*jump(u,n0)*(dS_vp + dS_hp)
+        )
+
+        if mu is not None:
+            eqn += dt*mu*inner(w,k)*inner(u,k)*dx
+        aeqn = lhs(eqn)
+        Leqn = rhs(eqn)
+
+        # Place to put result of u rho solver
+        self.urho = Function(M)
+
+        # Boundary conditions (assumes extruded mesh)
+        dim = M.sub(0).ufl_element().value_shape()[0]
+        bc = ("0.0",)*dim
+        bcs = [DirichletBC(M.sub(0), Expression(bc), "bottom"),
+               DirichletBC(M.sub(0), Expression(bc), "top")]
+
+        # Solver for u, rho
+        NEED TO BUILD THIS INTO A LAMBDA SOLVER
+
```

Comment:
Once the trace unknowns are added to the system, doing the Schur-complement to get the multiplier system is straightforward. Something like
```python
K0 = Tensor(beta*cp*dot(thetabar*w, n)*l0*(dS_vp + dS_hp))
K1 = Tensor(dl*dot(u,n0)*(dS_vp + dS_hp))
A = Tensor(aeqn) # It's already defined on the broken space at this point

S = assemble(K1 * A.inv * K0, bcs=...)
R = assemble(K1 * A.inv * Leqn)
lambdar = Function(Vtrace)
solve(S, lambdar, R, solver_parameters={...})
```